File: goofy-protocol-fis-be/src/main/java/com/masl/goofy_protocol_fis_be/crypto/SecretUtils.java
package com.masl.goofy_protocol_fis_be.crypto;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import java.security.spec.KeySpec;

public class SecretUtils {
    private static final Logger log = LoggerFactory.getLogger(SecretUtils.class);

    public static final String ENC_DELIMITER = ".";
    public static final int DEFAULT_ITERATIONS = 100_000;
    public static final String DEFAULT_DETERMINISTIC_SALT = "Goofy Protocol Default Salt";

    public static final int DEFAULT_HANDLE_ROOT_ITERATIONS = 300_000;
    public static final int DEFAULT_HANDLE_WORD_ITERATIONS = 100_000;
    public static final String DEFAULT_HANDLE_ROOT_SALT = "Goofy Protocol Root Salt";
    public static final String DEFAULT_HANDLE_WORD_SALT = "Goofy Protocol Derived Word Salt";

    public static byte[] symmSecretFromSecret(String secret, String salt, int size) {
        return symmSecretFromSecret(secret, salt, size, DEFAULT_ITERATIONS);
    }

    public static byte[] symmSecretFromSecret(String secret, String salt, int size, int iterations) {
        try {
            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            KeySpec spec = new PBEKeySpec(secret.toCharArray(), salt.getBytes(), iterations, size);
            return factory.generateSecret(spec).getEncoded();
        } catch (Exception e) {
            log.error("Error generating symmetric secret from secret: {}", e.getMessage(), e);
            return null;
        }
    }
}

File: goofy-protocol-fis-test-fe/libs/goofy-protocol-core/crypto/isolated/secret-utils.mjs
// TODO: CHECK
export const ENC_DELIMITER = ".";

export const DEFAULT_ITERATIONS = 100_000;
export const DEFAULT_DETERMINISTIC_SALT = "Goofy Protocol Default Salt";

export const DEFAULT_HANDLE_ROOT_ITERATIONS = 300_000;
export const DEFAULT_HANDLE_WORD_ITERATIONS = 50_000;
export const DEFAULT_HANDLE_ROOT_SALT = "Goofy Protocol Root Salt";
export const DEFAULT_HANDLE_WORD_SALT = "Goofy Protocol Derived Word Salt";

export async function symmSecretFromSecret(secret, salt, size, iterations = DEFAULT_ITERATIONS) {
    // WebCrypto PBKDF2
    const enc = new TextEncoder();
    const keyMaterial = await crypto.subtle.importKey(
        "raw",
        enc.encode(secret),
        { name: "PBKDF2" },
        false,
        ["deriveBits"]
    );

    const bits = await crypto.subtle.deriveBits(
        {
            name: "PBKDF2",
            hash: "SHA-256",
            salt: enc.encode(salt),
            iterations,
        },
        keyMaterial,
        size * 8
    );

    return new Uint8Array(bits);
}

export async function sha256(inputBytesOrString) {
    const bytes = typeof inputBytesOrString === "string"
        ? new TextEncoder().encode(inputBytesOrString)
        : inputBytesOrString;

    const digest = await crypto.subtle.digest("SHA-256", bytes);
    return new Uint8Array(digest);
}